File: FunctionalTools/Algorithms/TS.cpp
#include <bits/stdc++.h>

using namespace std;
typedef long long ll;
typedef long double ld;
const ll INF = numeric_limits<ll>::max();

#define all(x) x.begin(), x.end()
#define fastio ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);

// asume f es unimodal (primero crece luego decrece) en [l..r]
ll ternary_search(int l, int r, function<ll(int)> f) {
    while (r - l > 3) {
        int m1 = l + (r - l) / 3;
        int m2 = r - (r - l) / 3;
        if (f(m1) < f(m2))
            l = m1;
        else
            r = m2;
    }
    ll best = f(l);
    for (int i = l+1; i <= r; i++)
        best = max(best, f(i));
    return best;
}

int main() {
    fastio;
    int L, R;
    cin >> L >> R;
    // ejemplo: f(x) = -(x-3)^2 + 10  (unimodal)
    auto f = [&](int x){ return -(ll)(x-3)*(x-3) + 10; };
    ll ans = ternary_search(L, R, f);
    cout << ans << "\n";
    return 0;
}

File: FunctionalTools/Codes/TS.py
import sys
input = sys.stdin.readline

def ternary_search(l, r, f):
    while r - l > 3:
        m1 = l + (r - l) // 3
        m2 = r - (r - l) // 3
        if f(m1) < f(m2):
            l = m1
        else:
            r = m2
    best = f(l)
    for x in range(l+1, r+1):
        best = max(best, f(x))
    return best

L, R = map(int, input().split())
# ejemplo unimodal: f(x) = -(x-3)^2 + 10
def f(x):
    return -(x-3)*(x-3) + 10

print(ternary_search(L, R, f))
